Index: linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvc_ctrl.c
===================================================================
--- linux-ubilinux-4.4.13.orig/drivers/media/usb/uvc/uvc_ctrl.c	2017-04-13 18:33:59.585411511 -0700
+++ linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvc_ctrl.c	2017-04-13 18:34:37.608096172 -0700
@@ -1973,12 +1973,16 @@
 	list_for_each_entry(entity, &chain->entities, chain) {
 		unsigned int i;

-		if (UVC_ENTITY_TYPE(entity) != UVC_VC_EXTENSION_UNIT ||
-		    !uvc_entity_match_guid(entity, mapping->entity))
+		if (!uvc_entity_match_guid(entity, mapping->entity))
 			continue;

 		for (i = 0; i < entity->ncontrols; ++i) {
 			ctrl = &entity->controls[i];
+			if (ctrl->initialized &&
+			    ctrl->info.selector == mapping->selector) {
+				found = 1;
+				break;
+			}
 			if (ctrl->index == mapping->selector - 1) {
 				found = 1;
 				break;
Index: linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvc_driver.c
===================================================================
--- linux-ubilinux-4.4.13.orig/drivers/media/usb/uvc/uvc_driver.c	2017-04-13 18:33:59.585411511 -0700
+++ linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvc_driver.c	2017-04-13 18:34:37.608096172 -0700
@@ -42,6 +42,11 @@
  * Video formats
  */

+/* definition for MUX format */
+#ifndef V4L2_PIX_FMT_MUX
+#define V4L2_PIX_FMT_MUX    v4l2_fourcc('M', 'U', 'X', ' ') /* MUX stream */
+#endif
+
 static struct uvc_format_desc uvc_fmts[] = {
 	{
 		.name		= "YUV 4:2:2 (YUYV)",
@@ -168,6 +173,16 @@
 		.guid		= UVC_GUID_FORMAT_RW10,
 		.fcc		= V4L2_PIX_FMT_SRGGB10P,
 	},
+	{
+		.name		= "MPEG2 TS",
+		.guid		= UVC_GUID_FORMAT_MPEG,
+		.fcc		= V4L2_PIX_FMT_MPEG,
+	},
+	{
+		.name		= "MUX",
+		.guid		= UVC_GUID_FORMAT_MUX,
+		.fcc		= V4L2_PIX_FMT_MUX,
+	},
 };

 /* ------------------------------------------------------------------------
@@ -478,6 +493,33 @@
 		break;

 	case UVC_VS_FORMAT_MPEG2TS:
+		n = dev->uvc_version >= 0x0110 ? 23 : 7;
+		if (buflen < n) {
+			uvc_trace(UVC_TRACE_DESCR, "device %d videostreaming "
+			       "interface %d FORMAT error\n",
+			       dev->udev->devnum,
+			       alts->desc.bInterfaceNumber);
+			return -EINVAL;
+		}
+
+		strlcpy(format->name, "MPEG2 TS", sizeof format->name);
+		format->fcc = V4L2_PIX_FMT_MPEG;
+		format->flags = UVC_FMT_FLAG_COMPRESSED | UVC_FMT_FLAG_STREAM;
+		format->bpp = 0;
+		ftype = 0;
+
+		/* Create a dummy frame descriptor. */
+		frame = &format->frame[0];
+		memset(&format->frame[0], 0, sizeof format->frame[0]);
+		frame->bFrameIntervalType = 0;
+		frame->dwDefaultFrameInterval = 1;
+		frame->dwFrameInterval = *intervals;
+		*(*intervals)++ = 1;
+		*(*intervals)++ = 10000000;
+		*(*intervals)++ = 1;
+		format->nframes = 1;
+		break;
+
 	case UVC_VS_FORMAT_STREAM_BASED:
 		/* Not supported yet. */
 	default:
@@ -754,6 +796,14 @@
 			break;

 		case UVC_VS_FORMAT_MPEG2TS:
+			/* MPEG2TS format has no frame descriptor. We will create a
+			 * dummy frame descriptor with a dummy frame interval range.
+			 */
+			nformats++;
+			nframes++;
+			nintervals += 3;
+			break;
+
 		case UVC_VS_FORMAT_STREAM_BASED:
 			uvc_trace(UVC_TRACE_DESCR, "device %d videostreaming "
 				"interface %d FORMAT %u is not supported.\n",
@@ -805,6 +855,7 @@
 		switch (buffer[2]) {
 		case UVC_VS_FORMAT_UNCOMPRESSED:
 		case UVC_VS_FORMAT_MJPEG:
+		case UVC_VS_FORMAT_MPEG2TS:
 		case UVC_VS_FORMAT_DV:
 		case UVC_VS_FORMAT_FRAME_BASED:
 			format->frame = frame;
Index: linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvc_v4l2.c
===================================================================
--- linux-ubilinux-4.4.13.orig/drivers/media/usb/uvc/uvc_v4l2.c	2017-04-13 18:33:59.585411511 -0700
+++ linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvc_v4l2.c	2017-04-13 18:34:37.608096172 -0700
@@ -381,11 +381,6 @@

 	mutex_lock(&stream->mutex);

-	if (uvc_queue_streaming(&stream->queue)) {
-		mutex_unlock(&stream->mutex);
-		return -EBUSY;
-	}
-
 	probe = stream->ctrl;
 	probe.dwFrameInterval =
 		uvc_try_frame_interval(stream->cur_frame, interval);
@@ -398,6 +393,13 @@
 	}

 	stream->ctrl = probe;
+
+	ret = uvc_commit_video(stream, &stream->ctrl);
+	if (ret < 0) {
+		mutex_unlock(&stream->mutex);
+		return ret;
+	}
+
 	mutex_unlock(&stream->mutex);

 	/* Return the actual frame period. */
Index: linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvc_video.c
===================================================================
--- linux-ubilinux-4.4.13.orig/drivers/media/usb/uvc/uvc_video.c	2017-04-13 18:33:59.585411511 -0700
+++ linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvc_video.c	2017-04-13 18:34:37.608096172 -0700
@@ -359,7 +359,7 @@
 	return ret;
 }

-static int uvc_commit_video(struct uvc_streaming *stream,
+int uvc_commit_video(struct uvc_streaming *stream,
 			    struct uvc_streaming_control *probe)
 {
 	return uvc_set_video_ctrl(stream, probe, 0);
Index: linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvcvideo.h
===================================================================
--- linux-ubilinux-4.4.13.orig/drivers/media/usb/uvc/uvcvideo.h	2017-04-13 18:33:59.585411511 -0700
+++ linux-ubilinux-4.4.13/drivers/media/usb/uvc/uvcvideo.h	2017-04-13 18:34:37.608096172 -0700
@@ -115,6 +115,12 @@
 #define UVC_GUID_FORMAT_M420 \
 	{ 'M',  '4',  '2',  '0', 0x00, 0x00, 0x10, 0x00, \
 	 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}
+#define UVC_GUID_FORMAT_MPEG \
+	{ 'M',  'P',  'E',  'G', 0x00, 0x00, 0x10, 0x00, \
+	 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}
+#define UVC_GUID_FORMAT_MUX \
+	{ 'M',  'U',  'X',  0x00, 0x00, 0x00, 0x10, 0x00, \
+	 0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}

 #define UVC_GUID_FORMAT_H264 \
 	{ 'H',  '2',  '6',  '4', 0x00, 0x00, 0x10, 0x00, \
@@ -683,6 +689,8 @@
 extern int uvc_video_enable(struct uvc_streaming *stream, int enable);
 extern int uvc_probe_video(struct uvc_streaming *stream,
 		struct uvc_streaming_control *probe);
+extern int uvc_commit_video(struct uvc_streaming *stream,
+			    struct uvc_streaming_control *probe);
 extern int uvc_query_ctrl(struct uvc_device *dev, __u8 query, __u8 unit,
 		__u8 intfnum, __u8 cs, void *data, __u16 size);
 void uvc_video_clock_update(struct uvc_streaming *stream,
